### PR TITLE
[SPIR-V] Allow intrinsics with aggregate return type to reach GlobalISel

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
Two main goals of this PR are:
* to support "Arithmetic with Overflow" intrinsics, including the special case when those intrinsics are being generated by the CodeGenPrepare pass during translations with optimization;
* to redirect intrinsics with aggregate return type to be lowered via GlobalISel operations instead of SPIRV-specific unfolding/lowering (see https://github.com/llvm/llvm-project/pull/95012).

There is a new test case `llvm/test/CodeGen/SPIRV/passes/translate-aggregate-uaddo.ll` that describes and checks the general logics of the translation.

This PR continues a series of PRs aimed to identify and fix flaws in code emission, to improve pass rates for the mode with expensive checks set on (see https://github.com/llvm/llvm-project/pull/101732, https://github.com/llvm/llvm-project/pull/104104, https://github.com/llvm/llvm-project/pull/106966), having in mind the ultimate goal of proceeding towards the non-experimental status of SPIR-V Backend.

The reproducers are:

1) consider `llc -O3 -mtriple=spirv64-unknown-unknown ...` with:

```
define spir_func i32 @foo(i32 %a, ptr addrspace(4) %p) {
entry:
  br label %l1

l1:
  %e = phi i32 [ %a, %entry ], [ %i, %body ]
  %i = add nsw i32 %e, 1
  %fl = icmp eq i32 %i, 0
  br i1 %fl, label %exit, label %body

body:
  store i8 42, ptr addrspace(4) %p
  br label %l1

exit:
  ret i32 %i
}
```

2) consider `llc -O0 -mtriple=spirv64-unknown-unknown ...` with:

```
define spir_func i32 @foo(i32 %a, ptr addrspace(4) %p) {
entry:
  br label %l1

l1:                                               ; preds = %body, %entry
  %e = phi i32 [ %a, %entry ], [ %math, %body ]
  %0 = call { i32, i1 } @llvm.uadd.with.overflow.i32(i32 %e, i32 1)
  %math = extractvalue { i32, i1 } %0, 0
  %ov = extractvalue { i32, i1 } %0, 1
  br i1 %ov, label %exit, label %body

body:                                             ; preds = %l1
  store i8 42, ptr addrspace(4) %p, align 1
  br label %l1

exit:                                             ; preds = %l1
  ret i32 %math
}
```

### CODE REVIEW

**Comment by VyacheslavLevytskyy** [comment]
FYI @efriedma-quic

File: llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp [hunk 1]
```diff
@@ -1270,6 +1316,8 @@ Instruction *SPIRVEmitIntrinsics::visitInsertValueInst(InsertValueInst &I) {
 }
 
 Instruction *SPIRVEmitIntrinsics::visitExtractValueInst(ExtractValueInst &I) {
+  if (I.getAggregateOperand()->getType()->isAggregateType())
+    return &I;
```
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:1320]: This looks strange to me. I'm new to the spir-v backend, so that does not mean too much. Why would we not generate the spv_extractv intrinsic now? I would expect that in essentially all cases the aggregate operand has an aggregate type.

**Review by s-perron — COMMENTED** [review]
Looks okay with my limited knowledge, but I did not understand one part.

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp [hunk 2]
```diff
@@ -590,6 +593,29 @@ bool SPIRVInstructionSelector::spvSelect(Register ResVReg,
   case TargetOpcode::G_USUBSAT:
     return selectExtInst(ResVReg, ResType, I, CL::u_sub_sat);
 
+  case TargetOpcode::G_SADDO:
+    report_fatal_error(
+        "Selection of a signed addition with overflow is not implemented",
```
Comment (efriedma-quic) [inline llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:598]: Instead of report_fatal_error, can you allow this to fallback to the generic expansion code?

File: llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp (hunk 1 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:1320]: You are right, it's counterintuitive (unfortunately).

The brief answer is that SPIRV Backend used to get rid of aggregate data types to conform to IRTranslator expectations, and so visitExtractValueInst() as a rule observes non-aggregate operands which are changed to i32 during "prepare-functions". If we saw an aggregate here we can be sure that translation would fail: aggregates are being decomposed to multiple v-regs and this triggers https://github.com/llvm/llvm-project/blob/e762d4dac762a3fc27c6e251086b6645d7543bb2/llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp#L2798

So this check means that there's no sense to generate spv_extractv, because the required logic has already been implemented using this PR's way.

A good illustration is when we run smth like `llc -O0 -mtriple=spirv64-unknown-unknown llvm/test/CodeGen/SPIRV/instructions/call-complex-function.ll -o 1 -print-before=prepare-functions -print-after=prepare-functions -print-after=emit-intrinsics -filter-print-funcs=foo` to check info between passes:

```
*** IR Dump Before SPIRV prepare functions (prepare-functions) ***
define i32 @foo({ i32, i16 } %in, i64 %unused) {
  %first = extractvalue { i32, i16 } %in, 0
  %bar = call i32 @fun(i32 %first)
  ret i32 %bar
}
*** IR Dump After SPIRV prepare functions (prepare-functions) ***
define i32 @foo(i32 %in, i64 %unused) {
  %first = extractvalue i32 %in, 0
  %bar = call i32 @fun(i32 %first)
  ret i32 %bar
}
*** IR Dump After SPIRV emit intrinsics (emit-intrinsics) ***
define i32 @foo(i32 %in, i64 %unused) {
  %1 = call i32 @llvm.spv.track.constant.i32.i32(i32 0, metadata i32 0)
  %2 = call i32 (i32, ...) @llvm.spv.extractv.i32(i32 %in, i32 %1)
  call void @llvm.spv.assign.type.i32(i32 %2, metadata i32 poison)
  %bar = call i32 @fun(i32 %2)
  call void (i32, ...) @llvm.spv.assign.name.i32(i32 %bar, i32 7496034)
  call void @llvm.spv.assign.type.i32(i32 %bar, metadata i32 poison)
  ret i32 %bar
}
```

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp (hunk 2 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp]: thank you, done

File: llvm/test/CodeGen/SPIRV/optimizations/add-check-overflow.ll [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+; CHECK: %[[#]] = OpLabel
+; CHECK: OpReturnValue %[[Val]]
+; CHECK: OpFunctionEnd
+
+define spir_func i32 @foo(i32 %a, ptr addrspace(4) %p) {
+entry:
+  br label %l1
+
+l1:
+  %e = phi i32 [ %a, %entry ], [ %i, %body ]
```
[the remark this answers is not in the document]
Comment (VyacheslavLevytskyy) [reply llvm/test/CodeGen/SPIRV/optimizations/add-check-overflow.ll:49]: thank you, changed

File: llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp [hunk 4]
```diff
@@ -188,6 +188,40 @@ bool isConvergenceIntrinsic(const Instruction *I) {
          II->getIntrinsicID() == Intrinsic::experimental_convergence_loop ||
          II->getIntrinsicID() == Intrinsic::experimental_convergence_anchor;
 }
+
+bool isInternalNonVoidIntrinsic(const Value *I) {
+  if (const auto *II = dyn_cast<IntrinsicInst>(I))
+    switch (II->getIntrinsicID()) {
```
[the remark this answers is not in the document]
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:194]: Almost, but not exactly. This list doesn't include int_spv_gep, for example, because it returns ptr type. This list is to include internal intrinsics that returns a non-ptr value and so may potentially participate in fake_use emission.

[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:194]: In that case, isn't the return type fetchable from the CallBase parent class? Or am I missing something? (Seems like maintaining a list like that is bound to end up wrong)

File: llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp [hunk 5]
```diff
@@ -462,6 +468,25 @@ generateAssignInstrs(MachineFunction &MF, SPIRVGlobalRegistry *GR,
             Def->getOpcode() != SPIRV::ASSIGN_TYPE)
           insertAssignInstr(Reg, Ty, nullptr, GR, MIB, MF.getRegInfo());
         ToErase.push_back(&MI);
+      } else if (MIOp == TargetOpcode::FAKE_USE && MI.getNumOperands() > 0) {
+        MachineInstr *MdMI = MI.getPrevNode();
```
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp:472]: Here is it assumed that `spv_value_md` always precedes `FAKE_USE`. If `MdMI` is not `spy_value_md` should it still be added to `ToErase`?

File: llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp]: I think you are right. My motivation was to explore the topic further and widen applicability of this PR's way to pass info through passes, but we may extend/address this later if needed. Another motivation was that internal intrinsics are intended for quite different things, and it's hard to separate them as a class to use in conditions, but this again doesn't matter in case of this PR specifically.

So for goals of this PR it may be better indeed to be more general and don't list intrinsics explicitly, and I've changed the explicit list to `F->getName().starts_with("llvm.spv.")` to exclude internal SPIR-V backend intrinsics.

File: llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp [hunk 6]
```diff
@@ -462,6 +468,25 @@ generateAssignInstrs(MachineFunction &MF, SPIRVGlobalRegistry *GR,
...
+          for (unsigned I = 1, E = MI.getNumOperands(); I != E && Def; ++I)
+            if (getVRegDef(MRI, MI.getOperand(I).getReg()) != Def)
+              Def = nullptr;
+          if (Def) {
+            const MDNode *MD = MdMI->getOperand(1).getMetadata();
+            StringRef ValueName =
+                cast<MDString>(MD->getOperand(1))->getString();
+            const MDNode *TypeMD = cast<MDNode>(MD->getOperand(0));
+            Type *ValueTy = getMDOperandAsType(TypeMD, 0);
+            GR->addValueAttrs(Def, std::make_pair(ValueTy, ValueName));
```
[the remark this answers is not in the document]
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp]: very good point, I surprised that it's StringRef, my bad, fixed

File: llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp [hunk 7]
```diff
@@ -1101,6 +1130,71 @@ bool SPIRVInstructionSelector::selectFence(MachineInstr &I) const {
...
+bool SPIRVInstructionSelector::selectOverflowArith(Register ResVReg,
+                                                   const SPIRVType *ResType,
+                                                   MachineInstr &I,
+                                                   unsigned Opcode) const {
+  Type *ResTy = nullptr;
+  StringRef ResName;
+  if (!GR.findValueAttrs(&I, ResTy, ResName))
+    report_fatal_error(
+        "Not enough info to select the arithmetic with overflow instruction");
+  if (!ResTy->isStructTy())
```
[the remark this answers is not in the document]
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp]: Not very plausible to expect ResTy be nullptr when GR returns successfully found record, but sanity check is a good thing anyway, thank you for the suggestion, I've added `!ResTy`.

File: llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp (hunk 5 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp:472]: > Here is it assumed that spv_value_md always precedes FAKE_USE.

In my opinion it's the weakest point of the whole PR. However, it allows us to move forward and improve things. This approach works and serves as a useful compromise so far, and I'm going to think about this exactly problem further.

> If MdMI is not spy_value_md should it still be added to ToErase?

True, we should remove only `spv_value_md`. Fixed, thank you.

File: llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp]: I recalled my reasoning why StringRef is a safe choice here, and actually it looks like a safe choice, but I agree that `std:string` makes it independent on implementation details and allows to reason about lifetime and safety in simpler terms (the last thing I'd wish to doubt in this code base is about lifetimes).

**Review by michalpaszkowski — APPROVED** [review]
